### PR TITLE
Pre Simulation Token Parameter Modifiers

### LINKED ISSUE #497: Add before_open Simulation Instance Arg and token-based parameter modification
Proposing to add a "short-term" approach to allow editing values in an INP file.  

Many users have brought attention (#464, #439, #308, #277, #174, #167) to the fact that we have not exposed every parameter to modify.  

> Our development philosophy has generally considered pre-processing as a separate concern to enabling the simulation interaction as pyswmm is well known for.   

Given popular demand (and before ... maybe if... we build better coverage around the modifying parameters using something like `swmmio`), I believe we can meet in the middle with a new feature.  This feature would pre-process an INP file and give the user flexibility to modify tokenized-positions (rather than by column names as `swmmio` does it).  Longer-term, I think we need to use standard getter/setting section-parameter name-specific features and properly integrate `swmmio` into a preprocessor.  However short-term, I think a very basic work around would be amazingly useful. 

This makes for a much cleaner workflow and better enables optimization. 

Here is the workflow that I am proposing:

Basic INP Section (Input_file_name.inp)
```
[JUNCTIONS]
;;               Invert     Max.       Init.      Surcharge  Ponded
;;Name           Elev.      Depth      Depth      Depth      Area
;;-------------- ---------- ---------- ---------- ---------- ----------
J1               2.816      15         0          0          0
J10              30.579     15         0          0          0
J11              31.592     15         0          0          0
J12              32.606     15         0          0          0
```

(Input_file_name_pre_processed.inp)
```
[JUNCTIONS]
;;               Invert     Max.       Init.      Surcharge  Ponded
;;Name           Elev.      Depth      Depth      Depth      Area
;;-------------- ---------- ---------- ---------- ---------- ----------
J1               2.816      20         5          0          0
J10              32         15         0          0          0
J11              31.592     15         0          0          0
J12              32.606     15         0          0          0
```

```python
from pyswmm import Simulation

def set_inp_params():
    setting_inp_dict = \
        {
        'JUNCTIONS':
            [
                'J1': 
                    {2: 20, 3: 5},
                'J10':
                    {1: 32}
            ] }
    return setting_inp_dict

# This creates, pre-processes parameters, and runs "Input_file_name_pre_processed.inp"
with Simulation('Input_file_name.inp', before_open=set_inp_params) as sim:
    for step in sim:
        pass
```

### PR DESCRIPTION
**Pull request by bemcdonnell**
Closes #497

### CODE REVIEW

File: pyswmm/tests/test_simulation.py
Comment (abhiramm7) [inline pyswmm/tests/test_simulation.py:1]: we might need more tests to test the pre-config functionality. But if you all think it is too much effort for now, I am good :D

Comment (bemcdonnell) [reply pyswmm/tests/test_simulation.py:1]: @abhiramm7 Thanks for the review!  I think this feature is a step in the direction of building a fully parameter key name implementation.  From a product development standpoint, the issues that I closed above suggest there is interest in fully building out the key-type coverage on the INP file.  I think we should circle back in a future 2.x release and iron out the implementation of that.   With this PR, I would still like @karosc to have a look.  This case isn't about perfection, rather, ensuring we are opening up space for new habits to form for out users.

File: pyswmm/simulation.py [hunk 1]
```diff
@@ -818,3 +828,239 @@ def test_callback():
...
+        """
+        return self._filename_suffix
+
+    @filename_suffix.setter
+    def filename_suffix(self, suffix: str):
+        """"""
+        self._filename_suffix = suffix
+
+    def add_update_by_token(self, new_val, section: str, id: str,
+                            index: int, row_num=0):
```
Comment (karosc) [inline pyswmm/simulation.py:942]: Stylistically I take issue with the order of the arguments here. I think hierarchically, so it'd read better to me as written below. Also, putting all arguments on seperate lines makes future git diffs cleaner and less likely to have conflicts.

`
def add_update_by_token(
    self,
    section: str,
    id: str,
    index: int, 
    new_val, 
    row_num=0
):
`

**Comment by bemcdonnell** [comment]
@karosc Thanks for diving in.  I think we're ready for final approval.  Green lights and all the comments addressed.

**Review by abhiramm7 — APPROVED** [review]

**Review by karosc — APPROVED** [review]

**Pull request merged**